src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs

```diff
@@ -4329,6 +4329,584 @@ static void Main()
+                //         F((bool b) => { if (b) return new B1(); return new B2(); });
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "F").WithArguments("Program.F<T>(System.Func<bool, T>)").WithLocation(13, 9),
+                // (14,9): error CS0411: The type arguments for method 'Program.F<T>(Func<bool, T>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //         F((bool b) => { if (b) return new C1(); return new C2(); });
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "F").WithArguments("Program.F<T>(System.Func<bool, T>)").WithLocation(14, 9));
+        }
+
+        // As above but with explicit return type.
+        [Fact]
+        public void BestType_02()
+        {
+            var source =
+@"using System;
+class A { }
+class B1 : A { }
+class B2 : A { }
+interface I { }
+class C1 : I { }
+class C2 : I { }
+class Program
+{
+    static void F<T>(Func<bool, T> f) { Console.WriteLine(typeof(T)); }
+    static void Main()
+    {
+        F(A (bool b) => { if (b) return new B1(); return new B2(); });
+        F(I (bool b) => { if (b) return new C1(); return new C2(); });
+    }
+}";
+            CompileAndVerify(source, parseOptions: TestOptions.RegularPreview, expectedOutput:
+@"A
+I");
+        }
+
+        // CS4031 is not reported for async lambda in [SecurityCritical] type.
+        [Fact]
+        [WorkItem(54074, "https://github.com/dotnet/roslyn/issues/54074")]
+        public void SecurityCritical_AsyncLambda()
+        {
+            var source =
+@"using System;
+using System.Security;
+using System.Threading.Tasks;
+[SecurityCritical]
+class Program
+{
+    static void Main()
+    {
+        Func<Task> f = async () => await Task.Yield();
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        // CS4031 is not reported for async lambda in [SecurityCritical] type.
+        [Fact]
+        [WorkItem(54074, "https://github.com/dotnet/roslyn/issues/54074")]
+        public void SecurityCritical_AsyncLambda_AttributeArgument()
+        {
+            var source =
+@"using System;
+using System.Security;
+using System.Threading.Tasks;
+class A : Attribute
+{
+    internal A(int i) { }
+}
+[SecurityCritical]
+[A(F(async () => await Task.Yield()))]
+class Program
+{
+    internal static int F(Func<Task> f) => 0;
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (9,4): error CS0182: An attribute argument must be a constant expression, typeof expression or array creation expression of an attribute parameter type
+                // [A(F(async () => await Task.Yield()))]
+                Diagnostic(ErrorCode.ERR_BadAttributeArgument, "F(async () => await Task.Yield())").WithLocation(9, 4));
+        }
```

Comment:
Consider testing a type inference scenario with an explicit return type. 
`M(int (i) => i)` with `M<T>(Func<T, T> f)` would be interesting.
Or something more basic like: `M(int (o) => 1)` with `M<T>(Func<object, T> f)`.

Reply:
Added those tests, although currently type inference makes explicit type inferences from anonymous method parameter types but not return type.